### PR TITLE
Select between offsets

### LINKED ISSUE #469: [feature request] Add select until offset command
A function to select all byte values between 2 offsets would be nice for reverse engineering purposes. Such as extracting a binary file embedded in another.

### PR DESCRIPTION
**Pull request by IngilizAdam**
Implements a simple command that enables user to select byte values between 2 offsets.
Useful for extracting embedded files. 

closes #469

### CODE REVIEW

File: src/selectBetweenOffsets.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    startingInput.placeholder = "Enter offset to select from";
+    endingInput.placeholder = "Enter offset to select until";
+
+    messaging.sendEvent({ type: MessageType.StashDisplayedOffset });
+
+    let fromOffset: number | undefined;
+    let toOffset: number | undefined;
+    let accepted = false;
+
+    startingInput.onDidChangeValue(value => {
```
Comment (connor4312) [inline src/selectBetweenOffsets.ts:19]: It seems we duplicate a lot of logic between the starting and ending inputs, can it be pulled into a helper function?

File: src/selectBetweenOffsets.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
+import * as vscode from "vscode";
+import { ExtensionHostMessageHandler, MessageType } from "../shared/protocol";
+
+const addressRe = /^0x[a-f0-9]+$/i;
+const decimalRe = /^[0-9]+$/i;
+
+export const showSelectBetweenOffsets = (messaging: ExtensionHostMessageHandler): void => {
+    const startingInput = vscode.window.createInputBox();
```
Comment (connor4312) [inline src/selectBetweenOffsets.ts:8]: It'd be quite nice for the default value in the starting input to be the currently selected byte. The HexDocument has a `selectionState` that you can read from.

**Review by connor4312 — COMMENTED** [review]
A couple comments, thanks for the PR :)

File: src/selectBetweenOffsets.ts (hunk 2 above)
Comment (IngilizAdam) [reply src/selectBetweenOffsets.ts:8]: I was looking for that but failed. Thanks, I will change it

File: src/selectBetweenOffsets.ts (hunk 1 above)
Comment (IngilizAdam) [reply src/selectBetweenOffsets.ts:19]: Yes, working on it

File: src/selectBetweenOffsets.ts (hunk 2 above)
Comment (IngilizAdam) [reply src/selectBetweenOffsets.ts:9]: I couldn't access the active HexDocument. So I added a static variable holding that reference. Is this okay or is there a better way to access it? @connor4312

**Comment by IngilizAdam** [comment]
@microsoft-github-policy-service agree

File: src/selectBetweenOffsets.ts (hunk 1 above)
Comment (IngilizAdam) [reply src/selectBetweenOffsets.ts]: How is it now?

File: src/selectBetweenOffsets.ts (hunk 2 above)
Comment (connor4312) [reply src/selectBetweenOffsets.ts:9]: You can get it via the `activeDocument` on the `HexEditorRegistry` (created inside activate())

Comment (IngilizAdam) [reply src/selectBetweenOffsets.ts:9]: Ohh, how did I not see that?? Fixing it now

File: src/selectBetweenOffsets.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+            startingInput.validationMessage = "Offset must be provided as a decimal (12345) or hex (0x12345) address";
+            return;
+        }
+
+        startingInput.validationMessage = "";
+        if (inputOffset !== undefined) {
+            startingInput.validationMessage = "";
+            messaging.sendEvent({ type: MessageType.GoToOffset, offset: inputOffset });
+        }
+    }
```
Comment (connor4312) [inline src/selectBetweenOffsets.ts:80]: This function always affects the `startingInput`, even though it's used for the endingInput as well

File: src/selectBetweenOffsets.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+        if (fromOffset !== undefined && toOffset !== undefined) {
+            accepted = true;
+            messaging.sendEvent({ type: MessageType.SetFocusedByteRange, startingOffset: fromOffset, endingOffset: toOffset });
+        }
+    });
+
+    startingInput.show();
+
+    function hideHelper() {
```
Comment (connor4312) [inline src/selectBetweenOffsets.ts:57]: you'll want to make sure to `.dispose()` each input after their used.

Reworking this function to use some inner `function getOffset(options: { ... }): Promise<number>` would probably be easier to deal with

File: src/selectBetweenOffsets.ts (hunk 3 above)
Comment (IngilizAdam) [reply src/selectBetweenOffsets.ts:80]: My bad, thanks for noticing

File: src/selectBetweenOffsets.ts (hunk 4 above)
Comment (IngilizAdam) [reply src/selectBetweenOffsets.ts:57]: I thought `onDidHide` already disposes the `InputBox`. In `goToOffset.ts` file, there isn't a `.dispose()` for the input. And if I call `.dispose()` in `hideHelper` function, `hideHelper` gets called twice. Should I still call `.dispose()`?

Comment (connor4312) [reply src/selectBetweenOffsets.ts:57]: Yes, we should always call `dispose`. There are additional resources that get collected.

Comment (IngilizAdam) [reply src/selectBetweenOffsets.ts]: I did what you suggested and I think it turned out great. Let me know if there are any problems.

**Review by connor4312 — APPROVED** [review]
Lgtm, thanks for the PR!

**Review by sandy081 — APPROVED** [review]

**Pull request merged**